src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/SessionMessageProcessor.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.ServiceBus;
+using Microsoft.Azure.WebJobs.Host.Executors;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.ServiceBus
+{
+    public class SessionMessageProcessor 
+    {
+        public SessionMessageProcessor(ClientEntity clientEntity, SessionHandlerOptions sessionHandlerOptions)
+        {
+            ClientEntity = clientEntity ?? throw new ArgumentNullException(nameof(clientEntity));
+            SessionHandlerOptions = sessionHandlerOptions ?? throw new ArgumentNullException(nameof(sessionHandlerOptions));
+        }
+
+        public SessionHandlerOptions SessionHandlerOptions { get; }
+
+        /// <summary>
+        /// Gets or sets the <see cref="ClientEntity"/> that will be used by the <see cref="ClientEntity"/>.
+        /// </summary>
+        protected ClientEntity ClientEntity { get; set; }
+
+        /// <summary>
+        /// This method is called when there is a new message to process, before the job function is invoked.
+        /// This allows any preprocessing to take place on the message before processing begins.
+        /// </summary>
+        /// <param name="session">The session associated with the message.</param>
+        /// <param name="message">The message to process.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to use.</param>
+        /// <returns>A <see cref="Task"/> that returns true if the message processing should continue, false otherwise.</returns>
+        public virtual async Task<bool> BeginProcessingMessageAsync(IMessageSession session, Message message, CancellationToken cancellationToken)
```

Comment:
Why await here?

Reply:
This method is virtual so we want to provide async method for custom `SessionMessageProcessor`

Reply:
I think what I'm saying you don't need  to.

```
abstract class TestAsyncOverride {
      internal virtual Task<bool> ExecuteAsync() {
          return Task.FromResult(true);
      }
    }

    class TestAsync: TestAsyncOverride {
        internal async override Task<bool> ExecuteAsync() {
            await Task.Delay(1000);
            return await base.ExecuteAsync();
        }
    }
```

Reply:
Make sense